integration/src/test/scala/com/snowplowanalytics/snowplow/collectors/scalastream/intergation/CustomPathsSpec.scala

```diff
@@ -0,0 +1,76 @@
+ * distributed under the Apache License Version 2.0 is distributed on an "AS
+ * IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the Apache License Version 2.0 for the specific language
+ * governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.collectors.scalastream.intergation
+
+import cats.effect.IO
+import cats.effect.testing.specs2.CatsIO
+import cats.syntax.all._
+import com.snowplowanalytics.snowplow.collectors.scalastream.intergation.CustomPathsSpec._
+import com.snowplowanalytics.snowplow.collectors.scalastream.intergation.TestUtils.Http.Request.RequestType.Good
+import com.snowplowanalytics.snowplow.collectors.scalastream.intergation.TestUtils.{Base64, EventGenerator, Http}
+import com.snowplowanalytics.snowplow.eventgen.collector.Api
+import org.specs2.matcher.MatchResult
+import org.specs2.matcher.Matchers.beRight
+import org.specs2.matcher.MustMatchers.theValue
+import org.specs2.mutable.Specification
+import org.testcontainers.containers.{GenericContainer => JGenericContainer}
+import org.testcontainers.containers.output.{OutputFrame, WaitingConsumer}
+
+import java.util.concurrent.{TimeUnit, TimeoutException}
+import java.util.function.Predicate
+
+class CustomPathsSpec extends Specification with CatsIO {
+  "The collector" should {
+    "correctly translate mapped custom paths" in {
+      val collector    = Containers.collector("stdout", "custom-paths")
+      val requestStubs = EventGenerator.makeStubs(3, 3)
+      val paths        = List(Api("com.acme", "track"), Api("com.acme", "redirect"), Api("com.acme", "iglu"))
+
+      val resources = for {
+        collector  <- Containers.mkContainer[IO](collector)
+        httpClient <- Http.mkHttpClient[IO]
+      } yield (collector, httpClient)
+
+      resources.use {
+        case (collector, httpClient) =>
+          val collectorPort = Containers.getExposedPort(collector, 12345)
+          val requests = requestStubs
+            .zip(paths)
+            .map { case (stub, path) => Http.Request.setPath(stub, path) }
+            .map(Http.Request.make(_, collectorPort, Good))
+
+          Http.sendAll[IO](requests, httpClient).map(_ => matchLogs(collector))
+      }
+    }
+  }
+}
+
+object CustomPathsSpec {
+  def matchLogs(container: JGenericContainer[_]): MatchResult[Either[TimeoutException, Unit]] = {
+    val logConsumer = new WaitingConsumer
+    val p = new Predicate[OutputFrame] {
+      override def test(t: OutputFrame): Boolean = {
+        val decoded = Base64.decode(t.getUtf8String)
+
+        decoded.contains("/com.snowplowanalytics.snowplow/tp2") ||
+          decoded.contains("/r/tp2") ||
+          decoded.contains("/com.snowplowanalytics.iglu/v1")
+      }
+    }
+
+    container.followOutput(logConsumer)
+    Either.catchOnly[TimeoutException](logConsumer.waitUntil(p, 1, TimeUnit.SECONDS, 3)) must beRight
+  }
+}
```

Comment:
> in fact we want to maintain it exactly for the purposes of testing

Because in the old world it was very convenient to use it instead of having to deploy a pipeline. Now that we are setting up automated integration tests, it will remove the need to manually run the collector to test it, it will be fully automated.

> it makes the tests much lighter. If we always use Kinesis for example, it means we always need at least one localstack container; and we introduce complexity like test having to wait on each other (if writing to the same stream), cleaning up the stream after a test, etc.

We can have one `localstack` instance for all the tests, and we can have one stream per test. It's possible to specify the list of streams to create programmatically when starting `localstack`, thanks to [KINESIS_INITIALIZE_STREAMS](https://docs.localstack.cloud/user-guide/aws/kinesis/#configuration).

> there is no extra logic on top of the core collector service

[There is!](https://github.com/snowplow/stream-collector/blob/master/stdout/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/StdoutSink.scala#L28-L38)

> If the tests use a specific sink, like Kinesis, then we are making it harder to debug if an issue comes from the core collector logic, or from something Kinesis-specific.

- If we care about what the collector emits, then we care about the serialized event that is emitted. Are you able to reconstruct `CollectorPayload` from the logs?
- Kinesis sink has been working properly for years, that's very unlikely that a failed test comes from there. And if it does, well that's great that the tests found a bug there!

Reply:
Good points on using one localstack for multiple tests, and separate streams per test. You're right that it can really help streamline a setup that requires localstack. And in fact, I'll definitely want to use it everywhere where localstack is required.

But my point is rather that many tests do not require localstack at all. We don't gain anything by using kinesis collector and localstack over using stdout collector only. So why do the more complicated thing, even if we can streamline it?

As I see it, everything we can do with stdout collector only, we should. No reason to add more resources and increase the test's running time, if we're not gaining anything. To answer your explicit question: yes, we can get the serialized event and construct a `CollectorPayload` from it. It's not in the PR yet, but it will be in the next iteration.

On the question of extra logic, yes, there is some in the stdout sink but it's as simple as can be. There's no retry and backoff logic, for example; and there is anything complex like redirecting events from one sink to another, like in the kinesis collector with sqs buffer enabled. I definitely agree that we should also be testing those complex sinks too; I just want to keep thing separate. A tests about `Set-Cookie` headers doesn't need to sink events to a stream at all.

Reply:
> We don't gain anything by using kinesis collector and localstack over using stdout collector only. So why do the more complicated thing, even if we can streamline it?

If we're able to get `List[Event]` and `List[BadRow]` from the collector's output and don't use `contains` on the logs, I agree that we don't get so much from using `localstack`.

> increase the test's running time

We're talking about a few seconds. But agreed that there is no reason to add `localstack` if we don't gain anything from it.

> A tests about Set-Cookie headers doesn't need to sink events to a stream at all.

Don't we want to check the output for each test to make sure that events are emitted/not emitted as they should each time?

Reply:
> Don't we want to check the output for each test... ?

I don't think we need to. As long as we have robust tests that verify that events are sunk appropriately, I reckon we can focus the rest of the tests on a specific behaviour and not worry about writing data.